bitcoin/BRSyncManager.c

```diff
@@ -0,0 +1,1549 @@
+    BRPeerManagerRescan (manager->peerManager);
+}
+
+typedef struct {
+    BRPeerSyncManager manager;
+    BRTransaction *transaction;
+} SubmitTransactionInfo;
+
+static void
+BRPeerSyncManagerSubmit(BRPeerSyncManager manager,
+                        OwnershipGiven BRTransaction *transaction) {
+    SubmitTransactionInfo *info = malloc (sizeof (SubmitTransactionInfo));
+    info->manager = manager;
+    info->transaction = transaction;
+
+    BRPeerManagerPublishTx (manager->peerManager,
+                            transaction,
+                            info,
+                            _BRPeerSyncManagerTxPublished);
+}
+
+static void
+BRPeerSyncManagerTickTock(BRPeerSyncManager manager) {
+    double progress = BRPeerManagerSyncProgress (manager->peerManager, 0);
+    uint8_t needSyncEvent = progress < 1.0;
+    if (needSyncEvent) {
+        if (0 == pthread_mutex_lock (&manager->lock)) {
+            needSyncEvent &= manager->isConnected && manager->isFullScan;
+            pthread_mutex_unlock (&manager->lock);
+        } else {
+            assert (0);
+        }
+    }
+
+    if (needSyncEvent) {
+        manager->eventCallback (manager->eventContext,
+                                BRPeerSyncManagerAsSyncManager (manager),
+                                (BRSyncManagerEvent) {
+                                    SYNC_MANAGER_SYNC_PROGRESS,
+                                    { .syncProgress = { progress * 100 }}
+                                });
+    }
+}
+
+/// MARK: - Peer Manager Callbacks
+
+static void
+_BRPeerSyncManagerSaveBlocks (void *info,
+                              int replace,
+                              OwnershipKept BRMerkleBlock **blocks,
+                              size_t count) {
+    BRPeerSyncManager manager = (BRPeerSyncManager) info;
+    manager->eventCallback (manager->eventContext,
+                            BRPeerSyncManagerAsSyncManager (manager),
+                            (BRSyncManagerEvent) {
+                                replace ? SYNC_MANAGER_SET_BLOCKS : SYNC_MANAGER_ADD_BLOCKS,
+                                {
+                                    .blocks = { blocks, count }
+                                }
+                            });
+}
+
+static void
+_BRPeerSyncManagerSavePeers  (void *info,
+                              int replace,
+                              OwnershipKept const BRPeer *peers,
+                              size_t count) {
+    BRPeerSyncManager manager = (BRPeerSyncManager) info;
+    manager->eventCallback (manager->eventContext,
+                            BRPeerSyncManagerAsSyncManager (manager),
+                            (BRSyncManagerEvent) {
+                                replace ? SYNC_MANAGER_SET_PEERS : SYNC_MANAGER_ADD_PEERS,
+                                {
+                                    .peers = { (BRPeer *) peers, count }
+                                }
+                            });
+}
```

Comment:
These Bitcoin callbacks (here and others) are happening in the Bitcoin thread and the `eventCallback` is a) guaranteed not to save event state (like `peers` above) and generally b) not to block excessively?

Reply:
I preserved the existing behaviour here because you had expressed concerns about persistence in the past. The handling of this event could be moved to the event thread but I erred on the side of existing behaviour.

Reply:
Some of the events are handled with a `signal` and some are handled directly (like file service ones).  I think the file service ones must be handled directly - because the peers, blocks, transactions are all stack allocated in the Bitcoin code.

Reply:
They all have copy routines (i.e. peers and merkleblocks) so they could be moved to the event handler if you'd like.

Reply:
How about a comment in the handler code for those cases?  To the effect of "Peers may be stack allocated by the caller; we'll process inline". Similarly for blocks, transactions.

Reply:
Ok. I had put a `OwnershipKept` on `_BRWalletManagerSyncEvent` to make that clear, but I can add a comment to be more explicit.